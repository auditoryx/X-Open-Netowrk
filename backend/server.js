require("dotenv").config({ path: __dirname + "/.env" });

const express = require("express");
const cors = require("cors");
const mongoose = require("mongoose");

// Import rate limiting middleware
const { 
  authLimiter, 
  apiLimiter, 
  ddosProtection, 
  ipBlacklist 
} = require("./middleware/rateLimiting");

const app = express();

// Apply security middleware first
app.use(ipBlacklist);
app.use(ddosProtection);

app.use(express.json());
app.use(cors());

// Apply general API rate limiting
app.use('/api/', apiLimiter);

console.log("🔍 MONGO_URI:", process.env.MONGO_URI || "Not Found");

// ✅ Connect to MongoDB
mongoose.connect(process.env.MONGO_URI, { serverSelectionTimeoutMS: 5000 })
.then(() => console.log("✅ Connected to MongoDB"))
.catch(err => console.error("❌ MongoDB connection error:", err));

// ✅ Import & Use Routes
const serviceRoutes = require(__dirname + "/routes/services");
const authRoutes = require(__dirname + "/routes/auth");

// Apply stricter rate limiting to auth routes
app.use("/api/auth", authLimiter, authRoutes);
app.use("/api/services", serviceRoutes);

// Health check endpoint (no rate limiting)
app.get("/health", (req, res) => {
  res.json({ status: "healthy", timestamp: new Date().toISOString() });
});

const PORT = process.env.PORT || 5000;
app.listen(PORT, () => console.log(`🚀 Server running on port ${PORT}`));
