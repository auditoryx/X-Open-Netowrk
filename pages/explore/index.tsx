import { useState, useEffect } from 'react';
import { useRouter, useSearchParams } from 'next/navigation';
import MapToggle from '@/components/Explore/MapToggle';
import WorldMapView from '@/components/Explore/WorldMapView';
import { getCoordsFromCity } from '@/lib/utils/getCoordsFromCity';
import dynamic from 'next/dynamic';
import Link from 'next/link';
import { searchCreators } from '@/lib/firestore/searchCreators'; // Import the function to search creators

// Interface for creator
interface Creator {
  id: string;
  displayName: string;
  location?: string;
  profileImageUrl?: string;
  roles?: string[];
  genres?: string[];
  bio?: string;
  verified?: boolean;
}

export default function ExplorePage() {
  const searchParams = useSearchParams();
  const router = useRouter();
  
  const [view, setView] = useState<'grid' | 'map'>(
    searchParams.get('view') === 'map' ? 'map' : 'grid'
  );
  
  const [creators, setCreators] = useState<Creator[]>([]);
  const [loading, setLoading] = useState(true);
  const [searchTerm, setSearchTerm] = useState(searchParams.get('q') || '');
  const [selectedCreator, setSelectedCreator] = useState<Creator | null>(null);
  
  const [filters, setFilters] = useState({
    role: searchParams.get('role') || '',
    location: searchParams.get('location') || '',
    genre: searchParams.get('genre') ? searchParams.get('genre')!.split(',') : [],
  });

  // Load creators
  useEffect(() => {
    async function loadCreators() {
      setLoading(true);
      try {
        // Use the searchCreators function to get creators based on filters and search term
        const creatorsData = await searchCreators({
          searchTerm,
          role: filters.role,
          location: filters.location,
          genre: filters.genre.join(','),
          limit: 100
        });
        
        setCreators(creatorsData);
      } catch (error) {
        console.error('Error loading creators:', error);
      } finally {
        setLoading(false);
      }
    }
    
    loadCreators();
  }, [searchTerm, filters]);

  const handleCreatorClick = (creator: Creator) => {
    setSelectedCreator(creator);
    // Show a modal or navigate to creator profile
    router.push(`/profile/${creator.id}`);
  };

  const handleSearch = (term: string) => {
    setSearchTerm(term);
    
    // Update URL
    const params = new URLSearchParams(searchParams.toString());
    if (term) {
      params.set('q', term);
    } else {
      params.delete('q');
    }
    router.push(`/explore?${params.toString()}`);
  };

  const handleFilterChange = (name: string, value: any) => {
    setFilters(prev => ({
      ...prev,
      [name]: value
    }));
    
    // Update URL
    const params = new URLSearchParams(searchParams.toString());
    if (value && (Array.isArray(value) ? value.length > 0 : value !== '')) {
      params.set(name, Array.isArray(value) ? value.join(',') : value);
    } else {
      params.delete(name);
    }
    router.push(`/explore?${params.toString()}`);
  };

  return (
    <div className="min-h-screen bg-gray-50">
      <div className="max-w-7xl mx-auto px-4 sm:px-6 lg:px-8 py-8">
        <div className="flex flex-col md:flex-row justify-between items-start md:items-center mb-6">
          <div>
            <h1 className="text-3xl font-bold text-gray-900">Explore Creators</h1>
            <p className="mt-1 text-sm text-gray-500">
              Find and connect with talented creators from around the world
            </p>
          </div>
          
          <div className="mt-4 md:mt-0">
            <MapToggle view={view} setView={setView} />
          </div>
        </div>
        
        <div className="mb-6">
          <div className="flex flex-col md:flex-row gap-4">
            <div className="flex-1">
              <input
                type="text"
                placeholder="Search creators by name, role, location..."
                className="w-full p-3 rounded-lg border border-gray-300 focus:ring-2 focus:ring-blue-500 focus:border-transparent"
                value={searchTerm}
                onChange={(e) => handleSearch(e.target.value)}
              />
            </div>
            
            <div className="flex gap-4">
              <select
                className="px-3 py-2 border rounded-lg"
                value={filters.role}
                onChange={(e) => handleFilterChange('role', e.target.value)}
              >
                <option value="">All Roles</option>
                <option value="producer">Producer</option>
                <option value="singer">Singer</option>
                <option value="songwriter">Songwriter</option>
                <option value="engineer">Engineer</option>
                <option value="dj">DJ</option>
                <option value="instrumentalist">Instrumentalist</option>
              </select>
              
              <select
                className="px-3 py-2 border rounded-lg"
                value={filters.location}
                onChange={(e) => handleFilterChange('location', e.target.value)}
              >
                <option value="">All Locations</option>
                <option value="New York">New York</option>
                <option value="Los Angeles">Los Angeles</option>
                <option value="London">London</option>
                <option value="Tokyo">Tokyo</option>
                <option value="Berlin">Berlin</option>
                <option value="Paris">Paris</option>
              </select>
            </div>
          </div>
        </div>
        
        {loading ? (
          <div className="flex justify-center items-center py-12">
            <div className="animate-spin rounded-full h-12 w-12 border-t-2 border-b-2 border-gray-900"></div>
          </div>
        ) : creators.length === 0 ? (
          <div className="text-center py-12">
            <h3 className="text-lg font-medium text-gray-900">No creators found</h3>
            <p className="mt-2 text-sm text-gray-500">Try changing your filters or search term</p>
          </div>
        ) : view === 'map' ? (
          <WorldMapView creators={creators} onCreatorClick={handleCreatorClick} />
        ) : (
          <div className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-6">
            {creators.map((creator) => (
              <div 
                key={creator.id}
                className="bg-white rounded-lg shadow-md overflow-hidden transition-transform hover:scale-105"
              >
                <Link href={`/profile/${creator.id}`}>
                  <div className="cursor-pointer">
                    <div className="h-48 bg-gray-200 relative">
                      {creator.profileImageUrl ? (
                        <img 
                          src={creator.profileImageUrl} 
                          alt={creator.displayName} 
                          className="w-full h-full object-cover"
                        />
                      ) : (
                        <div className="w-full h-full flex items-center justify-center bg-gray-100">
                          <svg className="h-16 w-16 text-gray-400" fill="currentColor" viewBox="0 0 24 24">
                            <path d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z" />
                          </svg>
                        </div>
                      )}
                      
                      {creator.verified && (
                        <div className="absolute top-2 right-2 bg-blue-500 text-white text-xs font-bold px-2 py-1 rounded-full">
                          Verified
                        </div>
                      )}
                    </div>
                    
                    <div className="p-4">
                      <h3 className="font-bold text-lg">{creator.displayName}</h3>
                      
                      {creator.location && (
                        <p className="text-sm text-gray-500 flex items-center mt-1">
                          <svg className="h-4 w-4 mr-1" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M17.657 16.657L13.414 20.9a1.998 1.998 0 01-2.827 0l-4.244-4.243a8 8 0 1111.314 0z" />
                            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M15 11a3 3 0 11-6 0 3 3 0 016 0z" />
                          </svg>
                          {creator.location}
                        </p>
                      )}
                      
                      {creator.roles && creator.roles.length > 0 && (
                        <div className="flex flex-wrap gap-1 mt-2">
                          {creator.roles.map(role => (
                            <span key={role} className="px-2 py-1 bg-gray-100 text-gray-800 text-xs rounded-full">
                              {role}
                            </span>
                          ))}
                        </div>
                      )}
                      
                      {creator.bio && (
                        <p className="mt-2 text-sm text-gray-600 line-clamp-2">{creator.bio}</p>
                      )}
                    </div>
                  </div>
                </Link>
              </div>
            ))}
          </div>
        )}
      </div>
    </div>
  );
}
