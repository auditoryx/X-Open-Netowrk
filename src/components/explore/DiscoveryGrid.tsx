'use client';

import { useEffect } from 'react';
import { track } from '@/lib/analytics/track';
import { useInfiniteQuery } from '@tanstack/react-query';
import { getNextAvailable } from '@/lib/firestore/getNextAvailable';
import { SaveButton } from '@/components/profile/SaveButton';
import { getAverageRating } from '@/lib/reviews/getAverageRating';
import { getReviewCount } from '@/lib/reviews/getReviewCount';
import { useRouter } from 'next/navigation';
import { getProfileCompletion } from '@/lib/profile/getProfileCompletion';
import { PointsBadge } from '@/components/profile/PointsBadge';
import { Translate } from '@/i18n/Translate';
import GenreBadges from '@/components/explore/GenreBadges';
import VerifiedBadge from '@/components/ui/VerifiedBadge';

type Page = { results: any[]; nextCursor?: string };

export default function DiscoveryGrid({ filters }: { filters: any }) {
  const {
    data,
    fetchNextPage,
    hasNextPage,
    isFetchingNextPage,
    isLoading,
  } = useInfiniteQuery<Page>({
    queryKey: ['creators', filters],
    initialPageParam: undefined,
    queryFn: async ({ pageParam }) => {
      const params = new URLSearchParams({ limit: '20', ...filters });
      if (filters.availableNow) params.set('availableNow', '1');
      if (pageParam) params.append('cursor', pageParam as string);
      track('search', { ...filters, page: pageParam ?? 1 });
      const res = await fetch(`/api/search?${params.toString()}`);
      const json = await res.json();
      const withMeta = await Promise.all(
        json.results.map(async (c: any) => {
          const next = await getNextAvailable(c.uid);
          const rating = await getAverageRating(c.uid);
          const count = await getReviewCount(c.uid);
          const completion = getProfileCompletion(c);
          return { ...c, next, rating, count, completion };
        })
      );
      return { results: withMeta, nextCursor: json.nextCursor };
    },
    getNextPageParam: last => last.nextCursor ?? undefined,
  });

  useEffect(() => {
    const onScroll = () => {
      if (
        window.innerHeight + window.scrollY >=
          document.body.offsetHeight - 100 &&
        hasNextPage &&
        !isFetchingNextPage
      ) {
        fetchNextPage();
      }
    };
    window.addEventListener('scroll', onScroll);
  if (data && data.pages[0].results.length === 0) return (<p className="p-4 text-gray-400">No creators found.<\/p>);
  }, [hasNextPage, isFetchingNextPage, fetchNextPage]);
  const router = useRouter();

  if (isLoading)
  if (data && data.pages[0].results.length === 0) return (<p className="p-4 text-gray-400">No creators found.<\/p>);
      <div className="text-white">
        <Translate t="explore.loadingResults" />
      </div>
    );

  const creators = data?.pages.flatMap(p => p.results) ?? [];

  if (data && data.pages[0].results.length === 0) return (<p className="p-4 text-gray-400">No creators found.<\/p>);
    <div className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 gap-6">
      {creators.map((creator: any) => (
        <div
          key={creator.uid}
          className="border border-white p-4 rounded hover:bg-white hover:text-black transition"
        >
          <div className="flex justify-between items-center mb-2">
            <h2 className="text-lg font-bold">
              {creator.name || <Translate t="common.unnamed" />} {creator.isVerified && <VerifiedBadge />}
            </h2>
            <SaveButton providerId={creator.uid} />
          </div>
          {creator.rating !== null && (
            <p className="text-yellow-500 text-sm mb-1">
              ‚≠ê {creator.rating.toFixed(1)} / 5.0 ({creator.count})
            </p>
          )}
          <p className="text-sm text-gray-500 mb-2 line-clamp-2">
            {creator.bio || <Translate t="common.noBio" />}
          </p>
          <p className="text-xs text-blue-400 mb-1">üìä {creator.completion}% Profile Complete</p>
          { (creator.role === 'artist' || creator.role === 'producer') && (
            <>
              <GenreBadges genres={(creator.genres || []).slice(0,3)} />
              {creator.minBpm && creator.maxBpm && (
                <span className="bg-neutral-700 text-xs px-2 py-0.5 rounded-full mr-1">
                  {creator.minBpm}-{creator.maxBpm} BPM
                </span>
              )}
            </>
          ) }
          <PointsBadge points={creator.points} />
          <button
            className="border px-4 py-1 rounded text-sm"
            onClick={() => router.push(`/profile/${creator.uid}`)}
            aria-label={`${(<Translate t="common.viewProfile" />)} ${creator.name}` as unknown as string}
          >
            <Translate t="common.viewProfile" />
          </button>
        </div>
      ))}
    </div>
  );
}
