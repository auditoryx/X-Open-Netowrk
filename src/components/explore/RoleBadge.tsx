'use client';
import React from 'react';
import { roleBadges, RoleKey } from '@/constants/roleBadges';

export function RoleBadge({
  role,
  profile,
}: {
  role: string;
  profile: Record<string, any>;
}) {
  const cfg = roleBadges[role as RoleKey];
  if (!cfg) return null;
  const key = cfg.label.toLowerCase();
  const value = profile?.[key];
  if (value === undefined || value === null || value === '') return null;
  return React.createElement(
    'span',
    { className: 'inline-block text-xs text-gray-400 mb-1' },
    `${cfg.icon} ${value} ${cfg.label}`
  );
}
