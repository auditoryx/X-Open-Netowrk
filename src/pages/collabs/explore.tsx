import { useState, useEffect } from 'react';
import { useRouter } from 'next/router';
import Link from 'next/link';
import { useAuth } from '@/lib/hooks/useAuth';
import { 
  getCollabPackages, 
  getCollabPackagesWithFilters, 
  getFeaturedCollabPackages 
} from '@/src/lib/firestore/getCollabPackages';
import { CollabPackageCard } from '@/src/components/collab/CollabPackageCard';
import { 
  CollabPackage, 
  CollabPackageFilters, 
  formatPackagePrice 
} from '@/src/lib/types/CollabPackage';
import toast from 'react-hot-toast';
import { 
  MagnifyingGlassIcon, 
  FunnelIcon,
  PlusIcon,
  StarIcon,
  MapPinIcon,
  ClockIcon,
  CurrencyDollarIcon
} from '@heroicons/react/24/outline';

export default function ExploreCollabPackagesPage() {
  const { user } = useAuth();
  const router = useRouter();
  
  const [packages, setPackages] = useState<CollabPackage[]>([]);
  const [featuredPackages, setFeaturedPackages] = useState<CollabPackage[]>([]);
  const [loading, setLoading] = useState(true);
  const [searchTerm, setSearchTerm] = useState('');
  const [showFilters, setShowFilters] = useState(false);
  const [filters, setFilters] = useState<CollabPackageFilters>({});
  const [sortBy, setSortBy] = useState<'newest' | 'price_low' | 'price_high' | 'rating' | 'popular'>('newest');

  useEffect(() => {
    loadPackages();
    loadFeaturedPackages();
  }, [filters, sortBy]);

  const loadPackages = async () => {
    try {
      setLoading(true);
      let packagesData: CollabPackage[];
      
      if (Object.keys(filters).length > 0) {
        packagesData = await getCollabPackagesWithFilters(filters, 50);
      } else {
        packagesData = await getCollabPackages(50);
      }
      
      // Apply search filter
      if (searchTerm.trim()) {
        packagesData = packagesData.filter(pkg => 
          pkg.title.toLowerCase().includes(searchTerm.toLowerCase()) ||
          pkg.description.toLowerCase().includes(searchTerm.toLowerCase()) ||
          pkg.tags.some(tag => tag.toLowerCase().includes(searchTerm.toLowerCase()))
        );
      }
      
      // Apply sorting
      packagesData = sortPackages(packagesData);
      
      setPackages(packagesData);
    } catch (error) {
      console.error('Error loading packages:', error);
      toast.error('Failed to load packages');
    } finally {
      setLoading(false);
    }
  };

  const loadFeaturedPackages = async () => {
    try {
      const featuredData = await getFeaturedCollabPackages(6);
      setFeaturedPackages(featuredData);
    } catch (error) {
      console.error('Error loading featured packages:', error);
    }
  };

  const sortPackages = (packages: CollabPackage[]): CollabPackage[] => {
    const sorted = [...packages];
    
    switch (sortBy) {
      case 'newest':
        return sorted.sort((a, b) => {
          const dateA = a.createdAt?.toDate?.() || new Date(a.createdAt);
          const dateB = b.createdAt?.toDate?.() || new Date(b.createdAt);
          return dateB.getTime() - dateA.getTime();
        });
      case 'price_low':
        return sorted.sort((a, b) => a.totalPrice - b.totalPrice);
      case 'price_high':
        return sorted.sort((a, b) => b.totalPrice - a.totalPrice);
      case 'rating':
        return sorted.sort((a, b) => (b.rating || 0) - (a.rating || 0));
      case 'popular':
        return sorted.sort((a, b) => (b.bookingCount || 0) - (a.bookingCount || 0));
      default:
        return sorted;
    }
  };

  const handleFilterChange = (newFilters: Partial<CollabPackageFilters>) => {
    setFilters(prev => ({ ...prev, ...newFilters }));
  };

  const clearFilters = () => {
    setFilters({});
    setSearchTerm('');
  };

  const handleSearch = () => {
    loadPackages();
  };

  if (loading && packages.length === 0) {
    return (
      <div className="min-h-screen bg-gray-50 dark:bg-gray-900">
        <div className="max-w-7xl mx-auto px-4 sm:px-6 lg:px-8 py-8">
          <div className="animate-pulse">
            <div className="h-8 bg-gray-200 dark:bg-gray-700 rounded w-1/3 mb-4"></div>
            <div className="h-4 bg-gray-200 dark:bg-gray-700 rounded w-2/3 mb-8"></div>
            <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6">
              {Array.from({ length: 6 }).map((_, i) => (
                <div key={i} className="bg-white dark:bg-gray-800 p-6 rounded-lg shadow-md">
                  <div className="h-48 bg-gray-200 dark:bg-gray-700 rounded mb-4"></div>
                  <div className="h-4 bg-gray-200 dark:bg-gray-700 rounded w-3/4 mb-2"></div>
                  <div className="h-4 bg-gray-200 dark:bg-gray-700 rounded w-1/2"></div>
                </div>
              ))}
            </div>
          </div>
        </div>
      </div>
    );
  }

  return (
    <div className="min-h-screen bg-gray-50 dark:bg-gray-900">
      <div className="max-w-7xl mx-auto px-4 sm:px-6 lg:px-8 py-8">
        {/* Header */}
        <div className="mb-8">
          <div className="flex flex-col sm:flex-row sm:items-center sm:justify-between">
            <div>
              <h1 className="text-3xl font-bold text-gray-900 dark:text-white">
                Explore Collab Packages
              </h1>
              <p className="mt-2 text-gray-600 dark:text-gray-400">
                Discover and book collaborative packages with verified creators
              </p>
            </div>
            
            {user && (
              <Link
                href="/collabs/create"
                className="mt-4 sm:mt-0 inline-flex items-center px-4 py-2 bg-blue-600 hover:bg-blue-700 text-white rounded-lg font-medium transition-colors"
              >
                <PlusIcon className="w-5 h-5 mr-2" />
                Create Package
              </Link>
            )}
          </div>
        </div>

        {/* Featured Packages */}
        {featuredPackages.length > 0 && (
          <div className="mb-12">
            <div className="flex items-center mb-6">
              <StarIcon className="w-6 h-6 text-yellow-500 mr-2" />
              <h2 className="text-xl font-semibold text-gray-900 dark:text-white">
                Featured Packages
              </h2>
            </div>
            <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6">
              {featuredPackages.map((package_) => (
                <CollabPackageCard
                  key={package_.id}
                  package={package_}
                  onClick={() => router.push(`/collabs/${package_.id}`)}
                />
              ))}
            </div>
          </div>
        )}

        {/* Search and Filters */}
        <div className="mb-8">
          <div className="flex flex-col sm:flex-row gap-4 mb-4">
            {/* Search */}
            <div className="flex-1 relative">
              <MagnifyingGlassIcon className="absolute left-3 top-1/2 transform -translate-y-1/2 w-5 h-5 text-gray-400" />
              <input
                type="text"
                value={searchTerm}
                onChange={(e) => setSearchTerm(e.target.value)}
                onKeyPress={(e) => e.key === 'Enter' && handleSearch()}
                placeholder="Search packages, tags, or descriptions..."
                className="w-full pl-10 pr-4 py-2 border border-gray-300 dark:border-gray-600 rounded-lg focus:ring-2 focus:ring-blue-500 dark:bg-gray-700 dark:text-white"
              />
              <button
                onClick={handleSearch}
                className="absolute right-2 top-1/2 transform -translate-y-1/2 px-3 py-1 bg-blue-600 hover:bg-blue-700 text-white rounded text-sm transition-colors"
              >
                Search
              </button>
            </div>
            
            {/* Filter Toggle */}
            <button
              onClick={() => setShowFilters(!showFilters)}
              className="flex items-center px-4 py-2 border border-gray-300 dark:border-gray-600 rounded-lg hover:bg-gray-50 dark:hover:bg-gray-700 transition-colors"
            >
              <FunnelIcon className="w-5 h-5 mr-2" />
              Filters
            </button>
            
            {/* Sort */}
            <select
              value={sortBy}
              onChange={(e) => setSortBy(e.target.value as typeof sortBy)}
              className="px-4 py-2 border border-gray-300 dark:border-gray-600 rounded-lg focus:ring-2 focus:ring-blue-500 dark:bg-gray-700 dark:text-white"
            >
              <option value="newest">Newest</option>
              <option value="price_low">Price: Low to High</option>
              <option value="price_high">Price: High to Low</option>
              <option value="rating">Highest Rated</option>
              <option value="popular">Most Popular</option>
            </select>
          </div>

          {/* Filter Panel */}
          {showFilters && (
            <div className="bg-white dark:bg-gray-800 rounded-lg shadow-md p-6 border border-gray-200 dark:border-gray-700">
              <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-6">
                {/* Roles Filter */}
                <div>
                  <label className="block text-sm font-medium text-gray-700 dark:text-gray-300 mb-2">
                    Roles
                  </label>
                  <div className="space-y-2">
                    {['artist', 'producer', 'engineer', 'videographer', 'studio'].map((role) => (
                      <label key={role} className="flex items-center">
                        <input
                          type="checkbox"
                          checked={filters.roles?.includes(role as any) || false}
                          onChange={(e) => {
                            const currentRoles = filters.roles || [];
                            const newRoles = e.target.checked
                              ? [...currentRoles, role as any]
                              : currentRoles.filter(r => r !== role);
                            handleFilterChange({ roles: newRoles });
                          }}
                          className="mr-2"
                        />
                        <span className="text-sm text-gray-700 dark:text-gray-300 capitalize">
                          {role}
                        </span>
                      </label>
                    ))}
                  </div>
                </div>

                {/* Price Range */}
                <div>
                  <label className="block text-sm font-medium text-gray-700 dark:text-gray-300 mb-2">
                    Price Range
                  </label>
                  <div className="space-y-2">
                    <input
                      type="number"
                      placeholder="Min price"
                      value={filters.priceRange?.min || ''}
                      onChange={(e) => handleFilterChange({
                        priceRange: {
                          ...filters.priceRange,
                          min: e.target.value ? parseFloat(e.target.value) : undefined
                        }
                      })}
                      className="w-full px-3 py-1 border border-gray-300 dark:border-gray-600 rounded text-sm dark:bg-gray-700 dark:text-white"
                    />
                    <input
                      type="number"
                      placeholder="Max price"
                      value={filters.priceRange?.max || ''}
                      onChange={(e) => handleFilterChange({
                        priceRange: {
                          ...filters.priceRange,
                          max: e.target.value ? parseFloat(e.target.value) : undefined
                        }
                      })}
                      className="w-full px-3 py-1 border border-gray-300 dark:border-gray-600 rounded text-sm dark:bg-gray-700 dark:text-white"
                    />
                  </div>
                </div>

                {/* Package Type */}
                <div>
                  <label className="block text-sm font-medium text-gray-700 dark:text-gray-300 mb-2">
                    Package Type
                  </label>
                  <select
                    value={filters.packageType || ''}
                    onChange={(e) => handleFilterChange({
                      packageType: e.target.value as CollabPackage['packageType'] || undefined
                    })}
                    className="w-full px-3 py-1 border border-gray-300 dark:border-gray-600 rounded text-sm dark:bg-gray-700 dark:text-white"
                  >
                    <option value="">All Types</option>
                    <option value="studio_session">Studio Session</option>
                    <option value="live_performance">Live Performance</option>
                    <option value="video_production">Video Production</option>
                    <option value="custom">Custom</option>
                  </select>
                </div>

                {/* Actions */}
                <div className="flex flex-col justify-end">
                  <button
                    onClick={clearFilters}
                    className="w-full px-4 py-2 text-gray-600 dark:text-gray-400 hover:text-gray-900 dark:hover:text-white border border-gray-300 dark:border-gray-600 rounded-lg hover:bg-gray-50 dark:hover:bg-gray-700 transition-colors"
                  >
                    Clear Filters
                  </button>
                </div>
              </div>
            </div>
          )}
        </div>

        {/* Results Header */}
        <div className="flex items-center justify-between mb-6">
          <p className="text-gray-600 dark:text-gray-400">
            {packages.length} package{packages.length !== 1 ? 's' : ''} found
          </p>
          
          {Object.keys(filters).length > 0 && (
            <div className="flex items-center text-sm text-gray-500 dark:text-gray-400">
              <span>Filters applied</span>
              <button
                onClick={clearFilters}
                className="ml-2 text-blue-600 hover:text-blue-700 underline"
              >
                Clear
              </button>
            </div>
          )}
        </div>

        {/* Packages Grid */}
        {packages.length > 0 ? (
          <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6">
            {packages.map((package_) => (
              <CollabPackageCard
                key={package_.id}
                package={package_}
                onClick={() => router.push(`/collabs/${package_.id}`)}
              />
            ))}
          </div>
        ) : !loading ? (
          <div className="text-center py-12">
            <div className="max-w-md mx-auto">
              <div className="mb-4">
                <MagnifyingGlassIcon className="w-16 h-16 text-gray-400 mx-auto" />
              </div>
              <h3 className="text-lg font-medium text-gray-900 dark:text-white mb-2">
                No packages found
              </h3>
              <p className="text-gray-600 dark:text-gray-400 mb-6">
                Try adjusting your search criteria or filters to find what you're looking for.
              </p>
              {Object.keys(filters).length > 0 && (
                <button
                  onClick={clearFilters}
                  className="inline-flex items-center px-4 py-2 bg-blue-600 hover:bg-blue-700 text-white rounded-lg font-medium transition-colors"
                >
                  Clear Filters
                </button>
              )}
            </div>
          </div>
        ) : null}

        {/* Load More (if needed) */}
        {packages.length > 0 && packages.length % 50 === 0 && (
          <div className="text-center mt-12">
            <button
              onClick={loadPackages}
              className="px-6 py-3 bg-white dark:bg-gray-800 border border-gray-300 dark:border-gray-600 rounded-lg hover:bg-gray-50 dark:hover:bg-gray-700 text-gray-700 dark:text-gray-300 font-medium transition-colors"
            >
              Load More Packages
            </button>
          </div>
        )}
      </div>
    </div>
  );
}
