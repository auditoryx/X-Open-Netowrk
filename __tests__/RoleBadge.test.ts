import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import { RoleBadge } from '@/components/explore/RoleBadge';

describe('RoleBadge', () => {
  const profile = {
    rooms: 3,
    travel: 'Worldwide',
    genre: 'Hip Hop',
    daw: 'Ableton',
    mix: 'Analog',
  };

  (['studio','videographer','artist','producer','engineer'] as const).forEach(role => {
    test(`renders ${role} emoji`, () => {
      const html = renderToStaticMarkup(
        React.createElement(RoleBadge, { role, profile })
      );
      expect(html).toMatchSnapshot();
    });
  });
});
