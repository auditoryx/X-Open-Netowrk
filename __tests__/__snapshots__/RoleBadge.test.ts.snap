// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoleBadge renders artist emoji 1`] = `"<span class="inline-block text-xs text-gray-400 mb-1">🎤 Hip Hop Genre</span>"`;

exports[`RoleBadge renders engineer emoji 1`] = `"<span class="inline-block text-xs text-gray-400 mb-1">🎚️ Analog Mix</span>"`;

exports[`RoleBadge renders producer emoji 1`] = `"<span class="inline-block text-xs text-gray-400 mb-1">🎛️ Ableton DAW</span>"`;

exports[`RoleBadge renders studio emoji 1`] = `"<span class="inline-block text-xs text-gray-400 mb-1">🏢 3 Rooms</span>"`;

exports[`RoleBadge renders videographer emoji 1`] = `"<span class="inline-block text-xs text-gray-400 mb-1">🎥 Worldwide Travel</span>"`;
